File: LeetCode-2023/剑指Offer 47.礼物的最大价值.py
class Solution:
    def maxValue(self, grid: list[list[int]]) -> int:
        m, n = len(grid), len(grid[0])
        val = [([0] * (n + 1)) for _ in range(m + 1)]
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                val[i][j] = max(val[i - 1][j], val[i][j - 1]) + grid[i - 1][j - 1]
        return val[m][n]

class Solution:
    def maxValue(self, grid: list[list[int]]) -> int:
        m, n = len(grid), len(grid[0])
        val = [([0] * (n)) for _ in range(m)]
        for i in range(0, m):
            for j in range(0, n):
                if i > 0:
                    val[i][j] = max(val[i][j], val[i-1][j])
                if j > 0:
                    val[i][j] = max(val[i][j], val[i][j-1])
                val[i][j] += grid[i][j]
        return val[m-1][n-1]

File: LeetCode-2023/剑指Offer 47.礼物的最大价值.cpp
// Solution 1:
class Solution {
public:
    int maxValue(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<int>> dp(m, vector<int>(n));
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(i > 0) {
                    dp[i][j] = max(dp[i][j], dp[i-1][j]);
                }
                if(j > 0) {
                    dp[i][j] = max(dp[i][j], dp[i][j-1]);
                }
                dp[i][j] += grid[i][j];
            }
        }
        return dp[m-1][n-1];
    }
};

// Solution 2:
class Solution {
public:
    int maxValue(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<int>> dp(m+1, vector<int>(n+1));
        dp[0][0] = grid[0][0];
        for(int i = 1; i < m+1; i++) {
            for(int j = 1; j < n+1; j++) {
                dp[i][j] = max(dp[i-1][j], dp[i][j-1]) + grid[i-1][j-1];
            }
        }
        return dp[m][n];
    }
};